Allow implicitly convertible functions as overrides

jO-Osko opened the pull request:
Implements lesser restrictions for functions to be explicitly convertible. Adds (so far only for return values) support to allow implicitly convertible functions in overrides. Fixes #11624.

Until now, `f : A_1 -> B_1` was implicitly convertible to `g : A_2 -> B_2` if and only if: `A_1 == B_1` and `A_2 == B_2`. 

This lessens the restriction to: `A_2` is implicitly convertible to `A_1` (contravariant in arguments) and `B_1` is implicitly convertible to `B_2`. 

I changed the error message from `types differ` to `types are not compatible` to better reflect the semantics of the new checking.

jO-Osko commented:
Allowing the overriding with `isImplicitly` convertible check for argument types is a bit more involved as argument types are currently used during OverrideChecker to check for override candidates and thus currently require equality. I am investigating how to get around this.

chriseth commented:
Thanks for the PR!
As far as I can see, this changes two things in the language:
 - allow overrides to change return types
 - allow return types to be different for implicit conversions of function types

For this to be safe, we have to check that the function still uses the same internal encoding: There is no code that actual performs the implicit conversion of return values, so it only works if the types are represented in exactly the same way.

jO-Osko committed:
Allow override when return types are safely implicitly convertible

jO-Osko committed:
Add safe implicit conversion of function types

jO-Osko committed:
Update documentation on function types

jO-Osko commented:
Hi
I updated the code so that only types that can safely be converted (have the same memory representation) are considered in this case and only in return types due. 

Currently, this seems to include contract types (intended use in the issue) and conversion from `address payable` to `address` as this is only on the type level.

I also updated the`isImplicitlyConvertible` for `FunctionType` to correspond to changes in inheritance.

jO-Osko committed:
Style fixes

chriseth commented:
I'm sorry, but I still have a rather bad feeling about this, I'm not sure if we want to allow to skip a conversion at the type level instead of at the representation level.
